lib/runner.js

```diff
@@ -95,100 +135,55 @@ class Runner extends EventEmitter {
 		this.results.push(result);
 		this.emit('test', props);
 	}
-	_buildStats() {
+
+	buildStats() {
 		const stats = {
-			testCount: 0,
+			failCount: 0,
+			knownFailureCount: 0,
+			passCount: 0,
 			skipCount: 0,
+			testCount: 0,
 			todoCount: 0
 		};
 
-		this.results
-			.map(result => {
-				return result.result;
-			})
-			.filter(test => {
-				return test.metadata.type === 'test';
-			})
-			.forEach(test => {
+		for (const result of this.results) {
+			if (!result.passed) {
+				// Includes hooks
+				stats.failCount++;
+			}
+
+			const metadata = result.result.metadata;
+			if (metadata.type === 'test') {
 				stats.testCount++;
 
-				if (test.metadata.skipped) {
+				if (metadata.skipped) {
 					stats.skipCount++;
-				}
-
-				if (test.metadata.todo) {
+				} else if (metadata.todo) {
 					stats.todoCount++;
+				} else if (result.passed) {
+					if (metadata.failing) {
+						stats.knownFailureCount++;
+					} else {
+						stats.passCount++;
+					}
 				}
-			});
-
-		stats.failCount = this.results
-			.filter(result => {
-				return result.passed === false;
-			})
-			.length;
-
-		stats.knownFailureCount = this.results
-			.filter(result => {
-				return result.passed === true && result.result.metadata.failing;
-			})
-			.length;
-
-		stats.passCount = stats.testCount - stats.failCount - stats.skipCount - stats.todoCount;
+			}
+		}
 
 		return stats;
 	}
+
 	run(options) {
 		if (options.runOnlyExclusive && !this.tests.hasExclusive) {
 			return Promise.resolve(null);
 		}
 
-		this.tests.on('test', this._addTestResult);
-
 		this.hasStarted = true;
-
-		return Promise.resolve(this.tests.build().run()).then(this._buildStats);
+		this.tests.on('test', result => this.addTestResult(result));
+		return new Promise(resolve => {
```

Comment:
Ah `Promise` here is indeed from `bluebird`. I guess I'll rename it to `Bluebird.try()`.